text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+
+Relay currently forwards all envelopes to the next relay in chain the same.  There are however benefits by being able
+to route sessions to different relays than for instance transactions or errors.  Particularly for future projects such
+as session replays and profiling it would be nice to be able to route experimental features to a specific relay rather
+than having to roll out the experimental support across the entire cluster of relays.
+
+### Modularization in Relay
+
+Clear up envelope processor to make it easier to add new items types etc and make it easier to add new functionality.
+Today the massive envelope processor needs to be updated for every single item type added.  This opens many questions
+about how different items should work on different layers of relay, how they are routed etc.  As we are adding more
+item types this can turn more into a serverless function like internal API for adding item types that make it easier
+to land new items.
+
+### Relay Disk Buffer
+
+Relays currently buffer all data in memory which means that if they go down, we lose data.  Instead relay ideally buffers
+all its data additionally in a Kafka topic or on-disk storage that can make recovery from extended downtime possible.
+
+### Partial Grouping Awareness in Relay
+
+Today our grouping system really can only run in the tail end of the processing pipeline.  This makes it impossible for
+us to perform certain types of filtering or grouping decisions in relay.  There is a range of issues that could already
+be grouped quite accurately in Relay which would then permit us to do metrics extraction in relay or to perform efficent
+discarding.
+
+### Tail-based Sampling / Distributed Sampling Context Buffer
+
+Relay is currently unable to perform tail based sampling or consistent sampling if the dynamic sampling context changes.
+This could be remedied by extensive buffering or alternative approaches.  While a full tail based sampling approach is
+likely to be quite costly, there might be hybrid approaches possible.
+
+## Service Infrastructure
+
+These are changes to the Sentry architecture to better support the creation of independent services.
+
+### In-Monolith Service Compartmentalization
+
+The Sentry monolith currently houses a range of small services that are however clustered together.  For instance the
+processing pipeline really could be separated entirely from the rest of the System.  However the same is true for quite
+a lot of different parts of the product.  For instance certain parts of the user experience are sufficiently idependent
+in the UI code today already that they could turn into a separate service that just supplies UI views together with some
+APIs.  While it's unclear if the existing experiences are worth splitting up into new services, it's probably likely that
+we can at least structure their code so that we can reduce the total amount of pieces of code that need to be checked and
+tested against changes.
```

Comment:
Is it worthwhile being explicit that service compartmentalization doesn't imply splitting the monolith into multiple repos and running them as independent processes. Similarly, compartmentalizing the UI doesn't mean 'micro-frontends' that are stitched together.

My understanding is that we want to compartmentalize build and test dependencies so that CI and code review can be done more efficiently is that what you had in mind?

Reply:
Pretty much yes. I think the discussion about single vs multi repo is largely orthogonal to how things are organized. We have pretty bad tooling for both mono repo and multi repo at the moment and I think in the absence of a clear direction I would personally probably keep the status quo for now.

We can still have a separate conversation later if we want to move things more into one repo or into multiple repos.

Reply:
I feel we should clarify what we really want to achieve with a Service Infrastructure. I think different people have different ideas on the goal, and depending on the goal the solution is different. 
I heard about testability, isolating feature to reduce risk, isolating feature to reduce deployment cost (which would require independent stand alone services).

I would add another one which is architecture sanity (when everything depends on everything there is no architecture).
Do we have other reasons ?

Reply:
I think that covers it.

(For me one hidden reason about the service discussion is that if we acknowledge that we already operate services, we would do a better job articulating the operational requirements. Today all services we create are one-off creations which makes operating the entirety outside of the primary sentry.io installation challenging. Though that applies less to the monolith than other things we operate but which we might separate from the monolith in the future.)